propolis/src/dispatch/sync_tasks.rs

```diff
@@ -0,0 +1,401 @@
+use std::collections::BTreeMap;
+use std::io::Result;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Condvar, Mutex, MutexGuard};
+use std::thread::{Builder, JoinHandle};
+
+use crate::vcpu::*;
+
+use super::{DispCtx, SharedCtx};
+
+pub type WakeFn = dyn Fn(&DispCtx) + Send + 'static;
+pub type SyncFn = dyn FnOnce(&mut SyncCtx) + Send + 'static;
+
+#[derive(Eq, PartialEq, Ord, PartialOrd)]
+enum Ident {
+    Vcpu(usize),
+    Custom(String),
+}
+struct Detail {
+    join: Option<JoinHandle<()>>,
+    ctrl: Arc<WorkerCtrl>,
+    wake: Option<Box<WakeFn>>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum State {
+    Run,
+    WaitQuiesce,
+    WaitShutdown,
+    Quiesce,
+    Shutdown,
+}
+
+struct Inner {
+    state: State,
+    workers: BTreeMap<Ident, Detail>,
+}
+
+pub(super) struct SyncDispatch {
+    inner: Mutex<Inner>,
+    cv: Condvar,
+}
+
+impl SyncDispatch {
+    pub(super) fn new() -> Self {
+        Self {
+            inner: Mutex::new(Inner {
+                state: State::Quiesce,
+                workers: BTreeMap::new(),
+            }),
+            cv: Condvar::new(),
+        }
+    }
+
+    pub(super) fn spawn_vcpu(
+        &self,
+        shared: SharedCtx,
+        vcpu: VcpuHdl,
+        vcpu_fn: VcpuRunFunc,
+    ) {
+        let mut inner = self.inner.lock().unwrap();
+        let ctrl = WorkerCtrl::create_held();
+        let mut ctx = SyncCtx::for_worker(shared, Arc::clone(&ctrl));
+        let id = vcpu.cpuid() as usize;
+        let name = format!("vcpu-{}", id);
+        let hdl = Builder::new()
+            .name(name)
+            .spawn(move || {
+                // wait at dispatch hold point until start
+                if ctx.check_yield() {
+                    return;
+                }
+
+                vcpu_fn(vcpu, &mut ctx);
+            })
+            .unwrap();
+
+        inner.workers.insert(
+            Ident::Vcpu(id),
+            Detail {
+                join: Some(hdl),
+                ctrl,
+                wake: Some(Box::new(move |ctx: &DispCtx| {
+                    let _ = ctx.mctx.vcpu(id).barrier();
+                })),
+            },
+        );
+    }
+
+    /// Spawns a new dedicated worker thread named `name` which invokes
+    /// `func` on `data`.
+    ///
+    /// An optional `wake` function may be supplied, when invoked, this
+    /// function should trigger the worker to move to a barrier point.
+    pub fn spawn(
+        &self,
+        name: String,
+        func: Box<SyncFn>,
+        wake: Option<Box<WakeFn>>,
```

Comment:
Dunno how much you care - but for these functions operating on `Box<FunctionType>`, you could pretty easily use function-local generics and avoid boxing.

If we used the following:

```rust
pub type SyncFn = FnOnce(&mut SyncCtx) + Send + 'static;
```
We'd already be specifying the `Send` and `'static` constraints necessary to move these functions around.

https://doc.rust-lang.org/rust-by-example/fn/closures/input_parameters.html

Reply:
I wasn't wild about monomorphizing the innards of the dispatcher over every possible consumer.  The boxing wasn't a concern (at least from a perf standpoint) since a thread is being spawned for this anyways.  The async equivalent to this had to be generic (rather than `Box<dyn>`) because I simply couldn't figure out type machinations to accomplish it that pleased the compiler (or that didn't rely on the unstable async callback feature)